compiler/next/lib/resolution/resolution-types.cpp

```diff
@@ -17,11 +17,235 @@
+        formalIdx++;
+      }
+      assert(formalIdx < byFormalIdx.size());
+
+      // TODO: special handling for operators
+
+      FormalActual& entry = byFormalIdx[formalIdx];
+      entry.hasActual = true;
+      entry.actualIdx = actualIdx;
+      entry.actualType = actual.type;
+      actualIdxToFormalIdx[actualIdx] = formalIdx;
+    }
+    actualIdx++;
+  }
+
+  // Make sure that any remaining formals are matched by name
+  // or have a default value.
+  while (formalIdx < byFormalIdx.size()) {
+    if (byFormalIdx[formalIdx].actualIdx < 0 &&
+        untyped->formals[formalIdx]->initExpression() == nullptr) {
+      // formal was not provided and there is no default value
+      mappingIsValid = false;
+      failingFormalIdx = formalIdx;
+      return false;
+    }
+    formalIdx++;
+  }
+
+  mappingIsValid = true;
+  return true;
+}
+
+FormalActualMap FormalActualMap::build(const UntypedFnSignature* sig,
+                                       const CallInfo& call) {
+
+  FormalActualMap ret;
+  bool ok = ret.computeAlignment(sig, nullptr, call);
+  if (!ok) {
+    ret.mappingIsValid = false;
+  }
+  return ret;
+}
+
+FormalActualMap FormalActualMap::build(const TypedFnSignature* sig,
+                                       const CallInfo& call) {
+  FormalActualMap ret;
+  bool ok = ret.computeAlignment(sig->untypedSignature, sig, call);
+  if (!ok) {
+    ret.mappingIsValid = false;
+  }
+  return ret;
+}
+
+std::string TypedFnSignature::toString() const {
+  std::string ret = untypedSignature->functionId.toString();
+  ret += "(";
+  for (size_t i = 0; i < untypedSignature->formals.size(); i++) {
+    if (i != 0) {
+      ret += ", ";
+    }
+    ret += untypedSignature->formals[i]->name().c_str();
+    ret += " : ";
+    ret += formalTypes[i].toString();
+  }
+  ret += ")";
+
+  return ret;
+}
+
+
+void PoiInfo::accumulate(const PoiInfo& addPoiInfo) {
+  poiFnIdsUsed.insert(addPoiInfo.poiFnIdsUsed.begin(),
+                      addPoiInfo.poiFnIdsUsed.end());
+}
+
+// this is a resolved function
+// check is a not-yet-resolved function
+bool PoiInfo::canReuse(const PoiInfo& check) const {
+  assert(resolved && !check.resolved);
+
+  return false; // TODO -- consider function names etc -- see PR #16261
+}
+
+std::string ResolvedExpression::toString() const {
+  std::string ret;
+  ret += " : ";
+  ret += type.toString();
+  ret += " ; ";
+  if (!toId.isEmpty()) {
+    ret += " refers to ";
+    ret += toId.toString();
+  } else {
+    auto onlyFn = mostSpecific.only();
+    if (onlyFn) {
+      ret += " calls ";
+      ret += onlyFn->toString();
+    } else {
+      if (auto sig = mostSpecific.bestRef()) {
+        ret += " calls ref ";
+        ret += sig->toString();
+      }
+      if (auto sig = mostSpecific.bestConstRef()) {
+        ret += " calls const ref ";
+        ret += sig->toString();
+      }
+      if (auto sig = mostSpecific.bestValue()) {
+        ret += " calls value ";
+        ret += sig->toString();
```

Comment:
I don't think `sig->toString()` ends in a newline, so this could get very long and convoluted, right?

Reply:
Right, but it's sufficiently simple in the cases where I want to see it for debugging. These `toString` methods are currently basically like AST dumpers for debugging.